Controller

ivanivanovv13 opened the pull request.

rosenvalentinov21 committed:
Rewrite project

ivanivanovv13 committed:
Controller update:
-change java.version to 11 in pom.xml
-implemnt endpoints
-fix relation between objects
-add @Query anotation to method in LifterRepository

Commit:
Implemen logic for ading workout:
-add some DTO objects
-add some Anotations and change Anotations properties in the Controller
-add the logic for creating workout in the WorkoutService

Commit:
Add endpoint for creating and retrieving workouts:
-add Workout Controller
-add logic for creating workouts
-add a repository class for creating dynamic queries to the DB
-add some DTO objects

rosenvalentinov21 reviewed (changes requested):
There are some mistakes and things to consider for improvement, but overall i encountered good design and practices throughout the code, good and easy to follow logic in the WorkoutServiceImpl, consice and understandable naming, good object structure and real-world like object relationships.

ivanivanovv13 commented on src/main/java/XDirectionProject/service/impl/LifterServiceImpl.java:
```diff
@@ -25,13 +27,42 @@ public Lifter addLifter(Lifter lifter) {
+            if (!lifterRepository.existsLifterByEmail(updateDetails.getEmail()))
+                lifter.setEmail(updateDetails.getEmail());
+        }
+        if (updateDetails.getAge() != 0)
```
The point of the if and else statements is to check whether there is some value to search about. That's why we can't validate them because if they are not valid they won't pass, but in our case, it will create an SQL query without any criteria for searching.

rosenvalentinov21 replied:
Fine, but lets consider validating them where appropriate.

ivanivanovv13 replied:
I think there should be validation in the DTO.I planned it to be added in one of the next development cycles.

ivanivanovv13 commented on src/main/java/XDirectionProject/model/Person.java:
```diff
@@ -3,16 +3,16 @@
-import javax.persistence.Id;
+import javax.persistence.*;
 
 
```
JPA's column annotation is used to create the table properties in the Database. The nullable property of the Column annotation will assure that the properties in the database can't be null.

rosenvalentinov21 replied:
Okey, that`s fine, but if we add also the NonNull annotation provided by Lombok it will check for null value even before we persist to the database, which is more optimal than validation at DB level, my suggestion is to use both.

ivanivanovv13 replied:
I think validation of the object's properties should be in the DTO's. That way the rest request won't pass if there is a wrong input. I planned this to be added in one of  the next development cycles.

rosenvalentinov21 replied:
That is even better, i agree.

ivanivanovv13 committed:
Requsted change:
- change update method endpoint name
- make property Id of object TrainingLog private
- add field numberOfWorkout for object Workout
- change the name of the method that finds exercise by id

ivanivanovv13 commented on src/main/java/XDirectionProject/repository/ExerciseRepository.java:
```diff
@@ -0,0 +1,13 @@
+public interface ExerciseRepository extends JpaRepository<Exercise, Long> {
+
+    @Query(value = "SELECT * FROM exercise as e WHERE e.id=:id", nativeQuery = true)
+    Optional<Exercise> findExerciseByEmail(String id);
```
Done.

ivanivanovv13 commented on src/main/java/XDirectionProject/model/TrainingLog.java:
```diff
@@ -0,0 +1,27 @@
+public class TrainingLog {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    Long id;
```
Done.

ivanivanovv13 commented on src/main/java/XDirectionProject/controller/WorkoutController.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.ArrayList;
+
+@RestController
+@RequestMapping("/workout")
```
Done,

ivanivanovv13 commented on src/main/java/XDirectionProject/controller/LifterController.java:
```diff
@@ -20,21 +20,27 @@ public LifterController(LifterService lifterService) {
     }
 
+    @PostMapping("/update-lifter-details/{email}")
+    public ResponseEntity<LifterDTO> saveLifter(@PathVariable(name = "email") String email, @RequestBody LifterDTO newLifterDetails) {
```
Done.
